protocols/perf/src/bin/perf.rs

```diff
@@ -163,232 +163,48 @@ async fn client(
-                to_receive,
-            },
-        )?;
-    }
-
-    let mut finished = 0;
-    let start = Instant::now();
-
-    loop {
-        match swarm.next().await.unwrap() {
-            SwarmEvent::Behaviour(libp2p_perf::client::Event {
-                id: _,
-                result: Ok(_),
-            }) => {
-                finished += 1;
-
-                if finished == num {
-                    break;
-                }
-            }
-            e => panic!("{e:?}"),
-        }
-    }
-
-    let duration = start.elapsed().as_secs_f64();
-    let requests_per_second = num as f64 / duration;
-
-    info!(
-            "Finished: sent {num} {to_send} bytes requests with {to_receive} bytes response each within {duration:.2} s",
-        );
-    info!("- {requests_per_second:.2} req/s\n");
-
-    Ok(())
-}
-
-async fn sequential_connections_per_second(server_address: Multiaddr) -> Result<()> {
-    info!("start benchmark: sequential connections with single request per second");
-    let mut rounds = 0;
-    let to_send = 1;
-    let to_receive = 1;
-    let start = Instant::now();
-
-    let mut latency_connection_establishment = Vec::new();
-    let mut latency_connection_establishment_plus_request = Vec::new();
-
-    loop {
-        if start.elapsed() > Duration::from_secs(30) {
-            break;
-        }
-
-        let mut swarm = swarm().await?;
+    tokio::spawn(async move {
+        info!("start benchmark: custom");
 
         let start = Instant::now();
 
         let server_peer_id = connect(&mut swarm, server_address.clone()).await?;
 
-        latency_connection_establishment.push(start.elapsed().as_secs_f64());
-
-        perf(
-            &mut swarm,
-            server_peer_id,
-            RunParams {
-                to_send,
-                to_receive,
-            },
-        )
-        .await?;
-
-        latency_connection_establishment_plus_request.push(start.elapsed().as_secs_f64());
-        rounds += 1;
-    }
-
-    let duration = start.elapsed().as_secs_f64();
-
-    latency_connection_establishment.sort_by(|a, b| a.partial_cmp(b).unwrap());
-    latency_connection_establishment_plus_request.sort_by(|a, b| a.partial_cmp(b).unwrap());
-
-    let connection_establishment_95th = percentile(&latency_connection_establishment, 0.95);
-    let connection_establishment_plus_request_95th =
-        percentile(&latency_connection_establishment_plus_request, 0.95);
-
-    info!(
-            "Finished: established {rounds} connections with one {to_send} bytes request and one {to_receive} bytes response within {duration:.2} s",
+        perf(&mut swarm, server_peer_id, params).await?;
+
+        println!(
+            "{}",
+            serde_json::to_string(&BenchmarkResult {
+                upload_bytes: params.to_send,
+                download_bytes: params.to_receive,
+                r#type: "final".to_string(),
+                time_seconds: start.elapsed().as_secs_f64(),
+            })
```

Comment:
How about:

```rust
impl BenchmarkResult {
	fn final(params: RunParams, start: Instant) -> Self { }
}
```

Reply:
Given that each of these only have a single call site and given that it is within a single module only, I don't think it is worth the indirection.

Reply:
Hmm, it just makes it a bit easier to read this function because it less concerned with details. Don't feel strongly though.

Reply:
Will keep as is. I expect to iterate on this code rather sooner than later.